docs/develop/typescript/message-passing.mdx

```diff
@@ -196,107 +290,319 @@ await client.workflow.signalWithStart(yourWorkflow, {
+
+### Use `asyncio.Lock` to prevent concurrent handler execution {#control-handler-concurrency}
+
+Concurrent processes can interact in unpredictable ways.
+Incorrectly written [concurrent message-passing](/encyclopedia/workflow-message-passing#message-handler-concurrency) code may not work correctly when multiple handler instances run simultaneously.
+Here's an example of a pathological case:
+
+```typescript
+export async function myWorkflow(): Promise<MyWorkflowOutput> {
+  let x = 0;
+  let y = 0;
+  wf.setHandler(mySignal, async () => {
+    const data = await myActivity();
+    x = data.x;
+
+    // 🐛🐛 Bug!! If multiple instances of this handler are executing
+    // concurrently, then there may be times when the Workflow has x from one
+    // Activity execution and y from another.
+    await wf.sleep(500); // or await anything else
+
+    y = data.y;
+  });
+  ...
+}
+```
+
+Coordinating access using a lock (also known as a mutex) corrects this code.
+Locking makes sure that only one handler instance can execute a specific section of code at any given time:
+
+```typescript
+export async function myWorkflow(): Promise<MyWorkflowOutput> {
+  let x = 0;
+  let y = 0;
+  const lock = new Mutex();
+
+  wf.setHandler(mySignal, async () => {
+    await lock.runExclusive(async () => {
+      const data = await myActivity();
+      x = data.x;
+
+      // ✅ OK: node's event loop may switch now to a different handler
+      // execution, or to the main workflow function, but no other execution of
+      // this handler can run until this execution finishes.
+      await wf.sleep(500); // or await anything else
+
+      y = data.y;
+    });
+  });
+  return {
+    name: 'hello',
+  };
+}
+```
+
+
+## Message handler troubleshooting {#message-handler-troubleshooting}
+
+When sending a Signal, Update, or Query to a Workflow, your Client might encounter the following errors:
+
+- **The client can't contact the server**:
+  You'll receive a [`client.ServiceError`](https://typescript.temporal.io/api/classes/client.ServiceError) on which the `cause.code` attribute is [gRPC status code](https://grpc.io/docs/guides/status-codes/) 14 `UNAVAILABLE` (after some retries).
+
+- **The workflow does not exist**:
+  You'll receive an [`common.WorkflowNotFoundError`](https://typescript.temporal.io/api/classes/common.WorkflowNotFoundError) error.
+
+
+### Problems when sending a Signal {#signal-problems}
+
+When using Signal, the two errors described above are the only errors that will result from your requests.
+
+For Queries and Updates, the client waits for a response from the Worker and therefore additional errors may occur during the handler Execution by the Worker.
+
+### Problems when sending an Update {#update-problems}
+
+When working with Updates, you may encounter these problems:
+
+- **No Workflow Workers are polling the Task Queue**:
+  Your request will be retried by the SDK Client indefinitely.
```

Comment:
Should mention how to time-bound client side retries.

Reply:
Would you mind proposing some text? We can also do this in a follow-on.

Reply:
> Should mention how to time-bound client side retries.

Actually, I don't have time to fact check this, but is this even possible to do safely? I mean, the server itself must bail out of waiting before the client request effectively gets timed out, otherwise the client would think the update didn't went through due to time out, but the workflow may effectively have received the update… What's the semantic of grpc deadline in the case of the startUpdate call?

Reply:
We've decided to address this in follow-on work.